### PR TITLE
Feature/pok 124 static json file api

### PR DESCRIPTION
**Pull request by AmazingHooman**

### CODE REVIEW

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs [hunk 1]
```diff
@@ -100,7 +104,127 @@ public void UpdateReadModel(SpreadsheetImportReport importReport)
...
+                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+                WriteIndented = true
+            };
+
+            //
+            // Entities
+            // 
+            Console.WriteLine("generating json files for entities");
+            ICollection<EntityTypeReadModel> entities = _entityTypeMapper.MapFromDatabase(importReport).Keys;
+            File.WriteAllText("resources/entities.json", JsonSerializer.Serialize(entities, serializeOptions));
```
Comment (Finrod-Amandil) [inline PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs]: Nomenclature: Name this type of ressource consistently "EntityTypes", not "Entities". This resource resolves for every resource name, what *type* of entity that resource is.

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs [hunk 2]
```diff
@@ -100,7 +104,127 @@ public void UpdateReadModel(SpreadsheetImportReport importReport)
...
+
+            foreach (var entity in entities)
+            {
+                File.WriteAllText("resources/entities/" + entity.ResourceName + ".json", JsonSerializer.Serialize(entity, serializeOptions));
```
Comment (Finrod-Amandil) [inline PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs]: First thought was, that its a bit excessive to have a single JSON file for every resource name, but its nice and simple and still works with the existing API, so ok :)

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs [hunk 3]
```diff
@@ -100,7 +104,127 @@ public void UpdateReadModel(SpreadsheetImportReport importReport)
...
+            //
+            // learnable moves
+            //
+            Console.WriteLine("generating json files for learnable-moves");
+            ICollection<SimpleLearnableMoveReadModel> learnableMoves = _simpleLearnableMoveMapper.MapFromDatabase(importReport).Keys;
+            File.WriteAllText("resources/learnable-moves.json", JsonSerializer.Serialize(learnableMoves, serializeOptions));
+
+            foreach (var learnableMove in learnableMoves)
+            {
+                File.WriteAllText("resources/learnable-moves/" + learnableMove.MoveResourceName + ".json", JsonSerializer.Serialize(learnableMove, serializeOptions));
```
Comment (Finrod-Amandil) [inline PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs]: Yes this is the best solution I think. One file per move. For AND queries, the files for all moves that are queried are loaded and compared to find Pokémon that can learn all these moves.

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs [hunk 4]
```diff
@@ -100,7 +104,127 @@ public void UpdateReadModel(SpreadsheetImportReport importReport)
...
+            Console.WriteLine("generating json files for itemstats");
+            ICollection<ItemStatBoostPokemonReadModel> itemStats = _itemStatBoostPokemonMapper.MapFromDatabase(importReport).Keys;
+            File.WriteAllText("resources/itemstats.json", JsonSerializer.Serialize(itemStats, serializeOptions));
+
+            //
+            // learnable moves
+            //
+            Console.WriteLine("generating json files for learnable-moves");
+            ICollection<SimpleLearnableMoveReadModel> learnableMoves = _simpleLearnableMoveMapper.MapFromDatabase(importReport).Keys;
+            File.WriteAllText("resources/learnable-moves.json", JsonSerializer.Serialize(learnableMoves, serializeOptions));
```
Comment (Finrod-Amandil) [inline PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs]: Is this file necessary? One file per move should be enough...?

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs [hunk 5]
```diff
@@ -100,7 +104,127 @@ public void UpdateReadModel(SpreadsheetImportReport importReport)
...
+            ICollection<NatureReadModel> natures = _natureMapper.MapFromDatabase(importReport).Keys;
+            File.WriteAllText("resources/natures.json", JsonSerializer.Serialize(natures, serializeOptions));
+
+
+            //
+            // varieties
+            //
+            Console.WriteLine("generating json files for varieties");
+            ICollection<PokemonVarietyReadModel> varieties = _pokemonVarietyMapper.MapFromDatabase(importReport).Keys;
+            File.WriteAllText("resources/varieties.json", JsonSerializer.Serialize(varieties, serializeOptions));
```
Comment (Finrod-Amandil) [inline PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs:190]: Do you write the full set of details about Pokémon here? No reduced list for the Pokédex page?

File: PokeOneWeb/Frontend/pokeoneweb/src/app/core/components/stat-input/stat-input.component.ts [hunk 6]
```diff
@@ -188,10 +188,16 @@ export class StatInputComponent implements OnInit {
             return;
         }
         this.itemBoostService
-            .getItemStatBoostsForPokemon(this.model.pokemon.resourceName)
+            .getItemStats()
```
Comment (Finrod-Amandil) [inline PokeOneWeb/Frontend/pokeoneweb/src/app/core/components/stat-input/stat-input.component.ts]: getItemStatBoosts()

File: PokeOneWeb/Frontend/pokeoneweb/src/app/core/services/api/entity-type.service.ts [hunk 7]
```diff
@@ -17,6 +18,6 @@ export class EntityTypeService extends BaseService {
     }
 
     public getEntityTypeForPath(path: string): Observable<IEntityTypeModel> {
-        return this.http.get<IEntityTypeModel>(`${this.baseUrl}/getentitytypeforpath?path=${path}`, this.httpOptions);
+        return this.http.get<IEntityTypeModel>(`${environment.baseUrl}/entities/${path}.json`, this.httpOptions);
```
Comment (Finrod-Amandil) [inline PokeOneWeb/Frontend/pokeoneweb/src/app/core/services/api/entity-type.service.ts]: /entityTypes/

File: PokeOneWeb/Frontend/pokeoneweb/src/app/core/services/api/pokemon.service.ts [hunk 8]
```diff
@@ -42,12 +43,24 @@ export class PokemonService extends BaseService {
...
+        if (m32) results.push(this.http.get<IPokemonVarietyNameModel>(`${environment.baseUrl}/learnable-moves/${m32}.json`, this.httpOptions));
+        if (m33) results.push(this.http.get<IPokemonVarietyNameModel>(`${environment.baseUrl}/learnable-moves/${m33}.json`, this.httpOptions));
+        if (m34) results.push(this.http.get<IPokemonVarietyNameModel>(`${environment.baseUrl}/learnable-moves/${m34}.json`, this.httpOptions));
+        if (m41) results.push(this.http.get<IPokemonVarietyNameModel>(`${environment.baseUrl}/learnable-moves/${m41}.json`, this.httpOptions));
+        if (m42) results.push(this.http.get<IPokemonVarietyNameModel>(`${environment.baseUrl}/learnable-moves/${m42}.json`, this.httpOptions));
+        if (m43) results.push(this.http.get<IPokemonVarietyNameModel>(`${environment.baseUrl}/learnable-moves/${m43}.json`, this.httpOptions));
+        if (m44) results.push(this.http.get<IPokemonVarietyNameModel>(`${environment.baseUrl}/learnable-moves/${m44}.json`, this.httpOptions));
 
-        return this.http.get<IPokemonVarietyNameModel[]>(url, this.httpOptions);
+        return forkJoin(results);
```
Comment (Finrod-Amandil) [inline PokeOneWeb/Frontend/pokeoneweb/src/app/core/services/api/pokemon.service.ts]: Does this result in an "AND" query for all moves? Original query would be "(m11 OR m12 OR m13 OR m14) AND (...)". --> Reduce to 4 params and query "m11 AND m21 AND m31 AND m41"

**Review by Finrod-Amandil — CHANGES_REQUESTED** [review]

File: PokeOneWeb/Frontend/pokeoneweb/src/app/core/services/api/pokemon.service.ts (hunk 8 above)
Comment (AmazingHooman) [reply PokeOneWeb/Frontend/pokeoneweb/src/app/core/services/api/pokemon.service.ts]: How I understand the implementation of the current controller is that all the moves are an OR clause and my new implementation is also a big OR clause.

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs (hunk 2 above)
Comment (AmazingHooman) [reply PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs]: 🥇

File: PokeOneWeb/Frontend/pokeoneweb/src/app/core/components/stat-input/stat-input.component.ts (hunk 6 above)
Comment (AmazingHooman) [reply PokeOneWeb/Frontend/pokeoneweb/src/app/core/components/stat-input/stat-input.component.ts]: 👍

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs (hunk 5 above)
Comment (AmazingHooman) [reply PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs:190]: correct

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs (hunk 4 above)
Comment (AmazingHooman) [reply PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs]: Probably this file isn't required...

File: PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs (hunk 1 above)
Comment (AmazingHooman) [reply PokeOneWeb.DataSync/ReadModelUpdate/Impl/ReadModelUpdateService.cs]: 👍

**Comment by Finrod-Amandil** [comment]
To discuss:
- JSON files are not stored in location where WebApi project expects them...? --> Add content folder in WebApi to gitignore
- Pokémon search/filters not working
- Search Pokémon by learnset: Query structure (OR vs. AND)
- Calculation of bulk and total stat column
- 115MB data needs to be loaded to display Pokémon List
- Item.PlacedItems still contains ApplicationDbId in JSON files 
- Implement SonarQube feedback?

**Comment by Finrod-Amandil** [comment]
To discuss:
- JSON files are not stored in location where WebApi project expects them...?
- Pokémon search/filters not working
- Search Pokémon by learnset: Query structure (OR vs. AND)
- Calculation of bulk and total stat column
- 115MB data needs to be loaded to display Pokémon List
- Item.PlacedItems still contains ApplicationDbId in JSON files

**Comment by AmazingHooman** [comment]
> * JSON files are not stored in location where WebApi project expects them...? --> Add content folder in WebApi to gitignore
👍

**Comment by AmazingHooman** [comment]
Pokémon search/filters not working -> fixed
Calculation of bulk and total stat column -> fixed

**Comment by AmazingHooman** [comment]
115MB data needs to be loaded to display Pokémon List -> I will optimize it afterwards.

**Comment by AmazingHooman** [comment]
Item.PlacedItems still contains ApplicationDbId in JSON files -> fixed

**Review by Finrod-Amandil — APPROVED** [review]
